parsers/FO.py

```diff
@@ -44,48 +54,128 @@ def fetch_production(
+        target_date_local_to = target_date_local_to + timedelta(days=1)
 
+    url = f"https://www.sev.fo/api/elproduction/search?type=item&group=powerplant&from={target_date_local_from.strftime('%Y%m%d')}170000&to={target_date_local_to.strftime('%Y%m%d')}170000"
     ses = session or Session()
-    url = "https://www.sev.fo/api/realtimemap/now"
     response: Response = ses.get(url)
-    obj = response.json()
+    if not response.ok:
+        raise ParserException(
+            PARSER,
+            f"Exception when fetching production error code: {response.status_code}: {response.text}",
+            zone_key,
+        )
+
+    hourly_production_for_each_power_plant = response.json()
 
     production_breakdown_list = ProductionBreakdownList(logger)
-    production_mix = ProductionMix()
-    for key, value in obj.items():
-        if "Sum" in key or "Test" in key or "VnVand" in key:
-            # "VnVand" is the sum of hydro (Mýrarnar + Fossá + Heygar)
+
+    hourly_production_for_each_power_plant_by_hour = groupby(
+        hourly_production_for_each_power_plant, key=lambda x: x["date"]
+    )
+    for hour, group in hourly_production_for_each_power_plant_by_hour:
+        dt = datetime.fromisoformat(hour).replace(tzinfo=timezone.utc)
+
+        # filter out entries outside of time window of interest (which happens if DST)
+        end_of_day = target_date_utc_to + timedelta(days=1) - timedelta(microseconds=1)
+        if dt < target_date_utc_from or dt > end_of_day:
             continue
 
-        elif key.endswith(ZONE_KEY_TO_DATA_KEY[zone_key]):
-            # E stands for Energy
-            raw_generation_type: str = key.replace(ZONE_KEY_TO_DATA_KEY[zone_key], "")
-            generation_type = MAP_GENERATION.get(raw_generation_type)
-            if generation_type is None:
-                raise ParserException(
-                    PARSER, f"Unknown generation type: {raw_generation_type}", zone_key
+        production_mix = ProductionMix()
+
+        for power_plant_data in group:
+            # filter out entries for areas that are not relevant
+            area = power_plant_data["area"]
+            if (
+                zone_key == ZoneKey("FO-MI")
+                and area != "Main"
+                or zone_key == ZoneKey("FO-SI")
+                and area != "South"
+            ):
+                continue
+
+            production_mode = MAP_PRODUCTION_MODE[power_plant_data["motivepower"]]
+            production_mix.add_value(production_mode, power_plant_data["mwh"])
+
+        production_breakdown_list.append(
+            zoneKey=zone_key,
+            datetime=dt,
+            source="sev.fo",
+            production=production_mix,
+        )
+
+    # historical API has a ~2:25h delay, so supplement with realtime API data if applicable
+    # note that this will still leave a gap, but at least it will provide a realtime signal
+    if target_datetime > production_breakdown_list.events[-1].datetime:
```

Comment:
might not be obvious in the diff, but the contents of this block are essentially the old parser code

Reply:
I'd try and split these into two functions, that both return a ProductionBreakdownList, one for the historical data and one for the realtime data.

Then you can return it with `return ProductionBreakdownList.update(realtime, historical, logger).to_list()` which will ensure the historical data is always the one that is used even if there for some reason was a delay in the realtime data or it stopped updating.

Reply:
Good points! I've updated it but I feel like it's still not quite what you were thinking of so it might need another pass 😅  (I've still kept the conditional because I wanted to make sure we are still making a call to the realtime API only if it's needed e.g. no point if requesting data from the previous week)

Reply:
I'll take another look as soon as possible.